### PR TITLE
Improvements Droid Lifecycle

### PR DESCRIPTION
**Pull request by dimonovdd**
### Description of Change ###

When using [`Xamarin.Forms.Application`](https://github.com/xamarin/Xamarin.Forms/blob/5.0.0/Xamarin.Forms.Core/Application.cs) `Start`, `Resume` and etc. methods to track the lifecycle of an Android application, we got a problem that [`FormsAppCompatActivity`](https://github.com/xamarin/Xamarin.Forms/blob/842669423884bd8ebe3ce53e4c715c2d9d4af824/Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs#L39) lifecycle is tracked and not the application.

The changes proposed in this PR implement an analog of `AndroidX.Lifecycle.ProcessLifecycleOwner` and  `IComponentCallbacks`. 

Related with #606

### CODE REVIEW

**Comment by mattleibow** [comment]
Hmmm... What would be a use case for tracking the application stop? I understand you might want to save or something. But, won't the OS just kill apps in the background or does it wait for this to fire?

I recall reading somewhere that the reason for the missing events on application was because they don't get fired. But, since they added this in AndroidX, maybe this is changed...

If this is the case, then this is something we should do.

**Comment by mattleibow** [comment]
It seems the implementation just adds the lifecycle callbacks like we do and the events are for activities not the app: https://github.com/androidx/androidx/blob/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java

Maybe I am reading it wrong? Can't see any application events.

**Comment by Redth** [comment]
We do have delegates for the various Application lifecycle events: https://github.com/dotnet/maui/blob/main/src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs#L12-L16

We do implement most of them: https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/MauiApplication.cs#L35
We don't have `onTerminate()` but my understanding is this doesn't actually work well (or ever) in practice anyway.

We also implement a bunch of the Activity scope ones: https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/MauiAppCompatActivity.Lifecycle.cs
Though admittedly, we don't implement them _all_ yet.  PR's definitely welcome there! (ie: OnStop)

**Comment by dimonovdd** [comment]
> Maybe I am reading it wrong? Can't see any application events.

@mattleibow This is written in the [official android documentation](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner?hl=id):
 "Class that provides lifecycle for the whole application process."

**Comment by AmrAlSayed0** [comment]
I think `Activity` lifecycle events and `Application` lifecycle events are 2 different things. They may sometimes (mostly incase of single Activity) fire together but they're not always synonymous.

If an `Activity` corresponds to `Window` then it is not correct to assume it's lifecycle events applies to the whole application. This is wrong in 2 cases, multiple `Window`s/`Activity`s and background services.

I guess the current implementation in MAUI uses the `Activity`'s event because there are no corresponding events on the `Application` class (except for OnCreate). The `Xamarin.AndroidX.Lifecycle.Process` package provides those events and I think it will be a more accurate implementation.

**Comment by Redth** [comment]
@dimonovdd if you look at the actual source (find it online or decompile) for AndroidX's Lifecycle Process library you can see they simply use `registerActivityLifecycleCallbacks` to proxy the lifecycle events for activities and invoke the attributed methods.  It's just syntactic sugar for what we are already doing by surfacing the events from the activity itself.

As @mattleibow mentioned, there are no `Application` level lifecycle events raised by this androidx library.

There's no magic happening here, the androidx lifecycle lib is just a pleasant wrapper around the activity lifecycle callback api's.  We do not need this, we have the same information it provides, so it's better to avoid an additional dependency.  There's nothing more accurate about their implementation than ours.

Appreciate the PR and conversation around this :)

**Comment by dimonovdd** [comment]
@Redth Current implementation is good for one activity As @AmrAlSayed0  wrote.
I know it's not magic. If we don't want to add an extra dependency, then we can write it ourselves.
But why do this when there is a good solution?

>Yes, it collects events from all activity and processes them correctly for whole app

**Comment by Redth** [comment]
> @Redth Current implementation is good for one activity As @AmrAlSayed0 wrote.
> I know it's not magic. If we don't want to add an extra dependency, then we can write it ourselves.
> But why do this when there is a good solution?
> 
> > Yes, it collects events from all activity and processes them correctly for whole app

I'm confused, we already offer a way to subscribe to the `Application` level and/or the `Activity` level lifecycle events.  There's no need to write it yourself, just use the API's we are providing.

Yes the AndroidX Lifecycle library is _a_ solution, but it's unnecessary here and offers no tracking of `Application` specific lifecycle events.  I understand the documentation suggests it tracks processes, but I think the term _processes_ here is misunderstood.  I took a look at the code from the .jar file and it does not hook into `Application` at all, it simply uses the `registerActivityLifecycleCallbacks` API for the current Application and proxies the lifecycle callbacks to the attributed methods.

**Comment by dimonovdd** [comment]
> I'm confused

Probably it's my bad English.

1) When you subscribe via `registerActivityLifecycleCallbacks`, you get callbacks from all activities created in an app. and you will receive `OnStart` from all activity in  [`OnActivityStarted`](https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/MauiApplication.cs#L73)

2) `ProcessLifecycleOwner` does the same thing, but it counts how much activity [has started](https://github.com/androidx/androidx/blob/9c716662f43cbf115ab171de7da1f2155aaf08b3/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java#L110) and how much [has stopped](https://github.com/androidx/androidx/blob/9c716662f43cbf115ab171de7da1f2155aaf08b3/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java#L137). Here is such a ugly hack for tracking the application lifecycle. This is the single way for android.

I also think that `ProcessLifecycleOwner` code is just copied from stackoverflow.

**Comment by Redth** [comment]
> > I'm confused
> 
> Probably it's my bad English.
> 
> 1. When you subscribe via `registerActivityLifecycleCallbacks`, you get callbacks from all activities created in an app. and you will receive `OnStart` from all activity in  [`OnActivityStarted`](https://github.com/dotnet/maui/blob/main/src/Core/src/Platform/Android/MauiApplication.cs#L73)
> 2. `ProcessLifecycleOwner` does the same thing, but it counts how much activity [has started](https://github.com/androidx/androidx/blob/9c716662f43cbf115ab171de7da1f2155aaf08b3/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java#L110) and how much [has stopped](https://github.com/androidx/androidx/blob/9c716662f43cbf115ab171de7da1f2155aaf08b3/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java#L137). Here is such a ugly hack for tracking the application lifecycle. This is the single way for android.
> 
> I also think that `ProcessLifecycleOwner` code is just copied from stackoverflow.

So you're hoping to use the ref count of started/stopped activities to indicate when the app is closed?

**Comment by dimonovdd** [comment]
> So you're hoping to use the ref count of started/stopped activities to indicate when the app is closed?

Stopped means that an app is in the background.
I used `ProcessLifecycleOwner`  in my app and it worked cool. I had  a [`FormsAppCompatActivity`](https://github.com/xamarin/Xamarin.Forms/blob/842669423884bd8ebe3ce53e4c715c2d9d4af824/Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs#L39) and few others.

**Comment by dimonovdd** [comment]
I remembered another ugly hack:
```csharp
public class MainActivity : FormsAppCompatActivity
{
    private static App app;

    protected override void OnCreate(Bundle savedInstanceState)
    {
        base.OnCreate(savedInstanceState);
        Xamarin.Forms.Forms.Init(this, savedInstanceState);
        
        app ??= new App();
        LoadApplication(app);
    }
}
```
Activities can be recreated and we need to save the `App`  instance. And in this case, `ProcessLifecycleOwner` will also be useful, because `OnStart` and etc. methods will not be called several times

**Comment by AmrAlSayed0** [comment]
@dimonovdd Yes, I do that in my application too.

**Comment by mattleibow** [comment]
nd that is actually not a hack, but rather the way it should have worked from the start. An Activity.OnCreate is going to fire multiple times in an app's lifetime so you can't always just recreate the application.

I think this was bad from the start and we should have updated the templates.

**Comment by mattleibow** [comment]
We can also implement the activity counting in our implementation. I mean, no use bringing in a new package just to do that. Unless we are afraid of crazy changes.

**Comment by dimonovdd** [comment]
> We can also implement the activity counting in our implementation.

I can try to do this on the weekend. Oh should we use `Interlock` for it?

**Comment by mattleibow** [comment]
Hmmm... On the other hand, this looks like a very small package... Should we just use it?

Do we lose any events that we currently have? Seems pointless to have both our events and these events.

Then again, maybe it is easier if we just did it ourselves and not have to bring in the magic attributes and things.

You can look at the source for inspiration: https://github.com/androidx/androidx/blob/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java

I also see this random 700ms for this that makes me worry that they have their own timing issues :( https://github.com/androidx/androidx/blob/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java#L55


I am thinking it is better if we just implement some of this logic ourselves in our own system.

I think all the events come on the same thread since the app is the one firing it. So interlocked might not be needed and we can avoid all the locks.

**Comment by dimonovdd** [comment]
> I think all the events come on the same thread since the app is the one firing it. So interlocked might not be needed and we can avoid all the locks.

I also think so, but this is Android. It likes to give surprises

> I also see this random 700ms for this that makes me worry that they have their own timing issues :(

I may even have seen it on stackoverflow before AndroidX appeared. When an activity is replaced by another  `OnStop` method is called first and then `OnStart` method. 

[Official documentation says:](https://developer.android.com/reference/androidx/lifecycle/ProcessLifecycleOwner?hl=id)

>`Lifecycle.Event.ON_PAUSE`, `Lifecycle.Event.ON_STOP`, events will be dispatched with a **delay** after a last activity passed through them. This delay is long enough to guarantee that ProcessLifecycleOwner won't send any events if activities are destroyed and recreated due to a configuration change.

**Comment by Redth** [comment]
In this case I'm not so much worried about the size of the package but every androidx package we depend on has its own implications for version compatibility with other packages. This has caused enough grief on forms and in xamarin legacy that I want us to be careful when considering adding more. I don't feel the value is here for adding this one especially when the Java code for it is rather trivial and we can just make ours work the same way.

**Comment by dimonovdd** [comment]
I just copied the logic from [`ProcessLifecycleOwner`](https://github.com/androidx/androidx/blob/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.java).

We can also use `IComponentCallbacks` and `Activity Result API` to process other events from all activities

File: src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs [hunk 1]
```diff
@@ -11,10 +11,13 @@ public static class AndroidLifecycle
 		// Events called by the Application
 		public delegate void OnApplicationCreating(Application application);
 		public delegate void OnApplicationCreate(Application application);
-		public delegate void OnApplicationLowMemory(Application application);
-		public delegate void OnApplicationTrimMemory(Application application, TrimMemory level);
 		public delegate void OnApplicationConfigurationChanged(Application application, Configuration newConfig);
 
+		// Events called by IComponentCallbacks2
+		public delegate void OnConfigurationChanged(Configuration newConfig);
```
Comment (mattleibow) [inline src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs:17]: Does this still fire on activity events? Maybe rotation or keyboard things?

Comment (dimonovdd) [reply src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs:17]: Yes it works for a changing the orientation even if you don't add `ConfigChanges.Orientation` to `ConfigurationChanges` in your Activity. 


What should I do to have `OnConfigurationChanged` called by keyboard? It is never called in a activity, even if you use this attribute:

```csharp
[Activity(ConfigurationChanges = ConfigChanges.KeyboardHidden | ConfigChanges.Keyboard)]
```

**Comment by dimonovdd** [comment]
@mattleibow We can rollback the changes related to IComponentCallbacks2, I think it is not very useful

**Comment by Redth** [comment]
We'll look a bit more closely at this one this week, but I think this is generally the right idea.  I like the idea of moving the overrides from our activity, this makes it easier to have a different activity hook into it if necessary.

**Comment by dimonovdd** [comment]
@Redth Currently I can't help with this.

I changed my macbook to a new one with m1. And I'm having difficulty launching projects

https://developercommunity.visualstudio.com/t/SDK-not-found-MSBuildSdkExtras/1648593?from=email&space=41&q=sdk+not+found

**Comment by Redth** [comment]
No problem, we can finish getting it through to merged this week or next.  Appreciate the work on it so far :)

**Comment by dimonovdd** [comment]
@Redth HI
what do I need to do for us to merge this?

**Comment by jfversluis** [comment]
@Redth @mattleibow assigned you for now to this to help this move forward. Let me know if you won't have time in the near future and ideally nominate someone else 😬

File: src/Core/src/Platform/Android/MauiApplication.cs [hunk 2]
```diff
@@ -22,9 +19,9 @@ protected MauiApplication(IntPtr handle, JniHandleOwnership ownership) : base(ha
 
 		public override void OnCreate()
 		{
-			RegisterActivityLifecycleCallbacks(new ActivityLifecycleCallbacks());
-
 			var mauiApp = CreateMauiApp();
+			RegisterActivityLifecycleCallbacks(new ActivityLifecycleCallbacks(this, mauiApp.Services));
+			RegisterComponentCallbacks(new ApplicationComponentCallbacks(this, mauiApp.Services));
```
Comment (mattleibow) [inline src/Core/src/Platform/Android/MauiApplication.cs:24]: 2 things here:

1. The order has changed, so does this matter and will it affect users?
2. I like the concept of using the callbacks instead of overriding all the methods on the application.

Comment (mattleibow) [reply src/Core/src/Platform/Android/MauiApplication.cs:24]: I see we assumed services was around and if not we just skipped. So probably point 1. is not super valid actually.

File: src/Core/src/Platform/Android/ComponentCallbacks.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		protected IServiceProvider services;
+
+		protected ComponentCallbacks(IServiceProvider services)
+			=> this.services = services;
+
+		public abstract void OnConfigurationChanged(Configuration newConfig);
+
+		public virtual void OnLowMemory() { }
+
+		public virtual void OnTrimMemory(TrimMemory level) { }
```
Comment (mattleibow) [inline src/Core/src/Platform/Android/ComponentCallbacks.cs:53]: Maybe this should just be in the application level since that is where it only happens?

File: src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs [hunk 4]
```diff
@@ -11,9 +11,12 @@ public static class AndroidLifecycle
 		// Events called by the Application
 		public delegate void OnApplicationCreating(Application application);
 		public delegate void OnApplicationCreate(Application application);
+
+		// Events called by IComponentCallbacks2
+		public delegate void OnConfigurationChanged(Activity activity, Configuration newConfig);
+		public delegate void OnApplicationConfigurationChanged(Application application, Configuration newConfig);
 		public delegate void OnApplicationLowMemory(Application application);
 		public delegate void OnApplicationTrimMemory(Application application, TrimMemory level);
-		public delegate void OnApplicationConfigurationChanged(Application application, Configuration newConfig);
```
Comment (mattleibow) [inline src/Core/src/LifecycleEvents/Android/AndroidLifecycle.cs:16]: nit: no need to move this and rather group by activity and then application.

File: src/Core/src/Platform/Android/ComponentCallbacks.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		public ActivityComponentCallbacks(Activity activity, IServiceProvider services)
+			: base(services) => this.activity = new WeakReference<Activity>(activity);
+
+		public override void OnConfigurationChanged(Configuration newConfig)
+		{
+			if (!activity.TryGetTarget(out var target))
+				return;
+			MauiApplication.Current?.Application.ThemeChanged();
+			services.InvokeLifecycleEvents<AndroidLifecycle.OnConfigurationChanged>(del => del(target, newConfig));
+		}
```
Comment (mattleibow) [inline src/Core/src/Platform/Android/ComponentCallbacks.cs:39]: This is certainly a reason to make this PR land as we can move more out of the overrides in the maui activity as native embedding is not using that.

File: src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	class ActivityLifecycleCallbacks : Java.Lang.Object, Application.IActivityLifecycleCallbacks
+	{
+		int startedCounter = 0;
+		int resumedCounter = 0;
+		TimeSpan timeout = TimeSpan.FromMilliseconds(700);
+		Timer? timer;
+		bool pauseSent = true;
+		bool stopSent = true;
+		IServiceProvider services;
+		Application application;
```
Comment (mattleibow) [inline src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: This is the only part that scares me as we are using this counter that does not (maybe I am wrong) consider services and when applications have no UI. So, if I have an app that has push notifications that pop a toast say, will that get the application pause/resume?

I think the counting part may be best moved to the community toolkit and potentially moved to use the xplat windows or something so all platforms can use it. But even if not xplat, this type seems to mix both the lifecycle events and the activity counting into a single class.

Can we move the counting out of the object and into either a separate object or just use the android lifecycle events as this fires?

Basically, the counting seems very activity specific and we are basically syaing when all activities are minimized, the app is paused - Is that the case (with background services or notofocations/alarms)?

File: src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+		public void OnActivityDestroyed(Activity activity) =>
+			services?.InvokeLifecycleEvents<AndroidLifecycle.OnDestroy>(del => del(activity));
+
+		public void OnActivityStarted()
+		{
+			startedCounter++;
+			if (startedCounter != 1 || !stopSent)
+				return;
+			services?.InvokeLifecycleEvents<AndroidLifecycle.OnApplicationStart>(del => del(application));
```
Comment (mattleibow) [inline src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:67]: This is semi-interesting as this will fire once, and when the app is starting. So is this the same as the CreateMauiApp? This event seems to be tied to "application starts when the first activity is starting" and I am not sure that is correct as you could have an app that starts and does not show UI but still needs to run because reasons.

Maybe I am manufacturing cases. But for maui CORE, I would rather have less that is more accurate 100% than "mostly accurate".

File: src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs (hunk 6 above)
Comment (mattleibow) [reply src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: How does this work as well when the app is doing a StartActivityForResult and the dev settings is to kill the activity? Will the app report a stop and start while say picking photos? If it really is stopping the app then OK, but if it is just stopping the activity...

**Review by mattleibow — COMMENTED** [review]
Hello! Been a long time! And has changed a bit so I had to find where the comments applied to the current iteration.

I am on all sorts of fences here so will just list some points and hopefully it comes together near the end of this text box...

Starting from the bottom of the PR and working up...

... Overall, this PR looks nice and cleans up things, but has some extra points:
* I am scared of the activity counter in Core and I am unsure how this tracks to things that don't require or kill UI
* We need to not delete code but rather obsolete so we don't break customers.

After all the changes I think I like this PR, but I still have the fear of that counter and 700ms arb number that seems to "work". I would maybe like to move that out into the community toolkit unless we have "numbers" on what happens when various "app modes" are in play. 

@Redth You were also in this PR for a bit so wondering if the 2020 Jon is still talking to the 2023 Jon and how they feel about their comments and opinions on this.

File: src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs (hunk 6 above)
Comment (AmrAlSayed0) [reply src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: It's worth noting that android's implementation became much more involved since the last time this code was copied. https://github.com/androidx/androidx/tree/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle
Now it uses content providers and other initialization code to hook-in earlier than any activity.

The lifecycle.process package now depends on lifecycle.runtime package and androidx.startup.startup-runtime package

I guess that NOT taking a dependency also has it's costs like having to maintain the code 😅

File: src/Core/src/Platform/Android/MauiApplication.cs [hunk 8]
```diff
@@ -43,55 +40,10 @@ public override void OnCreate()
...
-			base.OnConfigurationChanged(newConfig);
-		}
-
 		public static MauiApplication Current { get; private set; } = null!;
 
 		public IServiceProvider Services { get; protected set; } = null!;
 
 		public IApplication Application { get; protected set; } = null!;
-
-		public class ActivityLifecycleCallbacks : Java.Lang.Object, IActivityLifecycleCallbacks
```
[the remark this answers is not in the document]
Comment (dimonovdd) [reply src/Core/src/Platform/Android/MauiApplication.cs:73]: It's been almost two years. It was done when the project was in the preview version

File: src/Core/src/Platform/Android/MauiApplication.cs (hunk 2 above)
Comment (dimonovdd) [reply src/Core/src/Platform/Android/MauiApplication.cs:24]: I think it happened during rebase

File: src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs (hunk 6 above)
Comment (dimonovdd) [reply src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: > It's worth noting that android's implementation became much more involved since the last time this code was copied.

Yes I would go back to depending. Google can improve this functionality

Comment (dimonovdd) [reply src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: It's not such a big package to be afraid of depending on it

Comment (dimonovdd) [reply src/Core/src/Platform/Android/ActivityLifecycleCallbacks.cs:18]: No, well, if you continue to measure the size of the application by  "Hello World", then this is of course a critical moment

File: src/Core/src/Core.csproj [hunk 9]
```diff
@@ -37,6 +37,7 @@
     <PackageReference Include="Xamarin.AndroidX.Navigation.Fragment" />
     <PackageReference Include="Xamarin.AndroidX.Navigation.Runtime" />
     <PackageReference Include="Xamarin.AndroidX.Navigation.Common" />
+    <PackageReference Include="Xamarin.AndroidX.Lifecycle.Process" Version="2.5.1.2"/>
```
Comment (dimonovdd) [inline src/Core/src/Core.csproj:40]: @mattleibow  Hi

I've redone it using `Xamarin.AndroidX.Lifecycle.Process`.  how should I install the package version correctly?
Is it also possible that you will do the review again? 

Thanks

Comment (Redth) [reply src/Core/src/Core.csproj:40]: You can add the package reference update + version here: https://github.com/dotnet/maui/blob/main/eng/AndroidX.targets

File: src/Core/src/Platform/Android/ComponentCallbacks.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+		readonly WeakReference<Activity> _activity;
+
+		public ActivityComponentCallbacks(Activity activity) =>
+			this._activity = new WeakReference<Activity>(activity);
+
+		public override void OnConfigurationChanged(Configuration newConfig)
+		{
+			if (!_activity.TryGetTarget(out var target))
+				return;
+			MauiApplication.Current?.Application.ThemeChanged();
```
Comment (Redth) [inline src/Core/src/Platform/Android/ComponentCallbacks.cs:37]: Does this belong here? Seems like this should maybe go in the host builder extension: https://github.com/dotnet/maui/blob/main/src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.Android.cs#L19

Or should this be called here at all?  What's the order of invocation and difference between the activity and application configuration changed being called?  Will they both be called?

File: src/Core/src/Platform/Android/MauiApplication.cs [hunk 11]
```diff
@@ -72,8 +52,11 @@ public override void OnConfigurationChanged(Configuration newConfig)
 
 		public class ActivityLifecycleCallbacks : Java.Lang.Object, IActivityLifecycleCallbacks
 		{
-			public void OnActivityCreated(Activity activity, Bundle? savedInstanceState) =>
+			public void OnActivityCreated(Activity activity, Bundle? savedInstanceState)
+			{
+				activity.RegisterComponentCallbacks(new ActivityComponentCallbacks(activity));
```
Comment (Redth) [inline src/Core/src/Platform/Android/MauiApplication.cs:57]: Should there be a matching UnregisterComponentCallbacks in the activity destroyed?

File: src/Core/src/Platform/Android/MauiApplication.cs [hunk 12]
```diff
@@ -23,6 +23,7 @@ protected MauiApplication(IntPtr handle, JniHandleOwnership ownership) : base(ha
 		public override void OnCreate()
 		{
 			RegisterActivityLifecycleCallbacks(new ActivityLifecycleCallbacks());
+			ProcessLifecycleOwner.Get().Lifecycle.AddObserver(new LifecycleObserver(this));
```
Comment (Redth) [inline src/Core/src/Platform/Android/MauiApplication.cs:26]: Should we have a matching RemoveObserver call for this?

**Comment by Redth** [comment]
Circling back to this one (finally, yes, sorry), unfortunately we are going to close this one for the time being.  Looking at this one again, it isn't aligned with our current set of priorities within the product.

I looked again at what the PR helped accomplish, and I think the main benefit is around tracking the app's backgrounded state based on ref counting the activity starts/stops.  There's also an interesting delay on surfacing the pause callback from an activity. 

Google has since migrated this code to kotlin, and the main implementation is here: https://github.com/androidx/androidx/blob/androidx-main/lifecycle/lifecycle-process/src/main/java/androidx/lifecycle/ProcessLifecycleOwner.kt

There's not much code here, so we could eventually decide to match the functionality in kotlin on our own (basically keeping ref counts ourselves, and surfacing new 'application' pseudo lifecycle events for stop/start.

If someone really needs to track this information today, they can definitely do so themselves with a little bit more code using our lifecycle builder apis.  Someone could even build a plugin around the concept and publish it to nuget if they felt so compelled.

I really appreciate the effort that went into the PR and conversation around it.  We've learned a bit about android lifecycles as a result.  Thank you!
